### ISSUE TITLE
Question about reproduction: were the reported results trained with the released script, and were tasks trained jointly or separately?

### ISSUE THREAD

**Issue by jingfengou** [ThinkMorph/ThinkMorph#18]
Hi, thanks for open-sourcing ThinkMorph.
I am trying to reproduce the results using the released code and evaluation pipeline, but I observed a significant discrepancy on some tasks, especially VSP / Spatial Navigation.
My main question is actually about the training setup used for the reported paper results.
What I want to confirm
Could u clarify:
1. Were the reported ThinkMorph results trained using the released training script in this repository, e.g. scripts/train_thinkmorph.sh?
2. Were the reported task results obtained from:
   - one single model jointly trained on the 4 released datasets, or
   - separate models trained per task/dataset?
This is especially important for understanding the reported VSP result.
My reproduction results
Using the released code and official evaluation pipeline, I obtained the following:

| Model / setting | VSP | VStar | VisPuzzle | BLINK-J | ChartQA | MMVP |
|---|---:|---:|---:|---:|---:|---:|
| Bagel (paper) | 0.83 | 55.49 | 35.00 | 67.33 | 62.05 | 70.33 |
| Bagel (my reproduction) | 11.67 | 7.85 | 9.50 | 29.33 | 46.00 | 9.33 |
| Interleave / ThinkMorph-style (paper) | 86.67 | 58.63 | 61.25 | 47.33 | 73.08 | 73.00 |
| Interleave / 4-task reproduction | 19.50 | 63.87 | 85.00 | 72.00 | 79.78 | 80.00 |

I also evaluated VSP by map size:

| Model | 3x3 (100) | 4x4 (100) | 5x5 (100) | 6x6 (100) | 7x7 (100) | 8x8 (100) | Overall |
|---|---:|---:|---:|---:|---:|---:|---:|
| bagel_base | 28.00 | 19.00 | 9.00 | 6.00 | 5.00 | 3.00 | 11.67 |
| thinkmorph_0002000 | 82.00 | 50.00 | 30.00 | 11.00 | 9.00 | 2.00 | 30.67 |
| thinkmorph_0008000 | 49.00 | 28.00 | 18.00 | 8.00 | 5.00 | 9.00 | 19.50 |

Why I am asking
The non-VSP tasks look mostly reasonable in my runs, and some are even higher than the reported numbers.  
However, VSP is much lower than the paper result, and performance drops sharply as the map size increases.
So I would like to verify whether the reported results correspond to:
- the released public training recipe,
- a single jointly trained model on the 4 datasets,
- or separate per-task training runs.
Any clarification would be very helpful. Thanks!

**Comment by jingfengou** [comment]
We have conducted experiments on the spatial navigation single-task using different environments (CUDA 12.1, FlashAttention 2.8.3 vs. CUDA 11.8, FA 2.7.3) with training durations of 2000 epochs and 3000 epochs, respectively. Surprisingly, the results exhibit significant discrepancies. Notably, the Cross-Entropy (CE) loss decreased much faster in the environment with CUDA 12.1 and FA 2.8.3. Below is the accuracy comparison table across different maze sizes and the corresponding loss plots.

| Model | 3x3 (100) | 4x4 (100) | 5x5 (100) | 6x6 (100) | 7x7 (100) | 8x8 (100) | Overall |
|---|---:|---:|---:|---:|---:|---:|---:|
| CUDA 12.1, FlashAttention 2.8.3 2000 epochs | 82.00 | 50.00 | 30.00 | 11.00 | 9.00 | 2.00 | 30.67 |
| CUDA 11.8, FA 2.7.3 3000 epochs | 66.00 | 71.00 | 78.00 | 72.00 | 50.00 | 30.00 | 61.17 |

<img width="2048" height="1195" alt="Image" src="https://github.com/user-attachments/assets/7d77a0dc-6936-42fb-992a-2dee004dbeff" />

The red line represents CUDA 11.8, FA 2.7.3 (3000 epochs), and the blue line represents CUDA 12.1, FlashAttention 2.8.3 (2000 epochs).

Could you please share your specific environment configurations as well?
Additionally, we would like to know whether you used model.safetensors or ema.safetensors during inference?

**Comment by jingfengou** [comment]
I noticed that you specifically saved the images in PNG format. I was wondering, does using different image formats have an effect on model training？
```python
def image_to_bytes(image):
    try:
        byte_stream = io.BytesIO()
        image.save(byte_stream, format='PNG')
        return byte_stream.getvalue()
    except Exception as e:
        print(f"[ERROR] Failed to process image: {e}")
        return None
```

**Comment by Kuvvius** [comment]
Hi, thanks for the detailed report. Quick answers below.

1. Yes, the released scripts are the actual ones we used. `train_thinkmorph.sh` trains the unified ThinkMorph model jointly on all four tasks for 8,000 steps. This is the ThinkMorph row in Table 3. `train_interleaved_reasoning.sh` runs the per-task fine-tune at 3,000 steps, one model per task. That is what each cell in Table 2 comes from. So VSP 86.67 is a model fine-tuned only on Spatial Navigation. Your 4-task joint number is from a different experiment. The gap is expected because VSP signal gets diluted in joint training. Your own numbers show this too. `thinkmorph_0008000` is worse on VSP than `thinkmorph_0002000`, because longer joint training drifts further from a VSP-specialized solution.

2. Step count is not the issue. The paper used 3,000 steps for every single-task fine-tune. lr 1e-5 cosine, max tokens 32,768 for interleaved, MSE and CE weights both 1. All listed in the appendix table. You also used 3,000 steps and got 71%. Training longer will not close the gap.

3. The most likely cause is data. Each VSP sample needs two images and two text segments. The two images are the input maze and a path-overlay reasoning image. The two text segments are an initial GPT-4.1 reasoning trace and a verification trace. The dataset covers grids from 3×3 to 6×6, 1,500 problems per size, 6,000 total. Missing any of these pieces will underfit even at the right step count. Also check that `num_used_data` in your yaml actually covers the full dataset. For evaluation, use the custom VSP prompt from the appendix. The default VLMEvalKit prompt gives a noticeably lower score on the same model.

4. PNG is intentional. VSP paths, arrows, and grid lines are pixel-sensitive. JPEG artifacts misalign the overlays just enough to hurt training. For natural-image tasks it matters less. For VSP keep PNG end-to-end.

5. Use `model.safetensors`, not `ema.safetensors`. That is what `inference.ipynb` loads (`load_checkpoint_and_dispatch(..., checkpoint="model.safetensors", dtype=torch.bfloat16)`). It is also what the HF release ships as the main file.

6. Our env matches `requirements.txt`: Python 3.10, torch 2.5.1+cu121, CUDA 12.1, flash-attn 2.5.8. The flash-attn line is commented out because there is no prebuilt wheel for torch 2.5.1. Install from source with `pip install flash-attn==2.5.8 --no-build-isolation`. Different flash-attn versions can shift per-step loss values from kernel non-determinism. But the causal mask code path is identical across 2.5.x, 2.7.x, and 2.8.x. The version is very unlikely to explain a 15-point VSP gap. Pinning 2.5.8 just removes a variable.

**Comment by jingfengou** [comment]
Thank you so much for taking the time to provide such a detailed and thorough response! I really appreciate the effort you put into breaking everything down so clearly; it is incredibly helpful.

Following your suggestions, I will double-check my data, hyperparameters, and other configurations. I plan to re-run the experiments and try to replicate the results within this specific environment. Hopefully, this will resolve the issue.

Thanks again for your amazing support and dedication to this project!
